File: practicas/practica_2/Deepspace_Ruby_P2/lib/CardDeck.rb
#encoding:utf-8

# File: CardDeck.rb
# Author: Profesor

module Deepspace

# Brief: Clase que representa un mazo de cartas de objetos
class CardDeck 
  
  # Constructor sin parámetros
  def initialize()
    @cards=Array.new()
    @ready=false
    @count=0
  end
  
  public
  
  # Brief: Inserta una nueva carta en el mazo
  # Param t: Carta a insertar
  def add (t)
    if !@ready then
      @cards << t
    end
  end
  
  # Brief: Devuelve la siguiente carta del mazo
  # Return: Siguiente carta del mazo
  def next ()
    if (!@ready) then
      @ready=true;
      shuffle();
    end
      
    card=@cards.delete_at(0);
    @cards.push(card);
    
    @count+=1
    if (@count==@cards.size()) then
      shuffle();
      @count=0;
    end
      
    return card.class.newCopy(card)
  end
  
  # Brief: Comprueba si se han barajado las cartas
  # Return: True si se han bajarado las cartas y False en otro caso
  def justShuffled()
    return (@count==0)
  end

  private
  
  # Brief: Baraja la colección de cartas disponibles
  def shuffle()
    @cards.shuffle!
  end
  
  # Brief: Función para representar una instancia de la clase CardDeck
  #        en un string
  # Return s: String que representa una instancia de la clase CardDeck
  def to_s
    getUIversion().to_s
  end
  
end # Class

end # Module

if $0 == __FILE__ then
  class TestCard
    attr_accessor :a, :b
    def initialize (a,b=nil)
      @a=a
      if b != nil then
        @b = Array.new(b)
      else 
        @b=Array.new()
        @b << 1
        @b << 2
        @b << 3
      end
    end
    
    def self.newCopy(tc)
      new(tc.a, tc.b)
    end
    
  end
  
  test=Deepspace::CardDeck.new
  test.add(TestCard.new(1))
  test.add(TestCard.new(2))
  test.add(TestCard.new(3))
  test.add(TestCard.new(4))
  test.add(TestCard.new(5))
  for i in 0..15 do
    c = test.next
    puts c.a
    puts "[ #{c.b[0]}  ]"
    c.a +=10
    c.b.delete_at(0)
  end
end

File: practicas/practica_2/deepspace_Java_P2/src/main/java/deepspace/CardDeck.java
/**
 * @file CardDeck.java
 * @author Profesor
 */
package deepspace;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @brief Clase parametrizable que representa un mazo de cartas de objetos de
 *        tipo T
 * @param <T> : Objeto del que se realizará el mazo de cartas
 */
class CardDeck<T> {    
    private ArrayList<T> cards=new ArrayList<>();
    private boolean ready;
    private int count;
    
    /**
     * @brief Constructor sin parámetros
     */
    CardDeck() {
        ready=false;
        count=0;
    }
    
    /**
     * @brief Inserta una nueva carta en el mazo
     * @param t Carta a insertar
     */
    public void add(T t) {
        if (!ready)
            cards.add(t);
    }
    
    /**
     * @brief Devuelve la siguiente carta del mazo
     * @return Siguiente carta del mazo
     */
    public T next() {
        if (!ready) {
            ready=true;
            shuffle();
        }

        T card=cards.remove(0);
        cards.add(card);
        
        count++;
        if (count==cards.size()) {
            shuffle();
            count=0;
        }
        
        T t=null;
        try {
            Class<T> clazz = (Class<T>) card.getClass();
            Constructor<T> c= clazz.getDeclaredConstructor(clazz);
            t = c.newInstance( card );
        } catch (ReflectiveOperationException ex) {
            Logger.getLogger(CardDeck.class.getName()).log(Level.SEVERE, null, ex);
        }
        return t;
        
        //return card;
    }
    
    /**
     * @brief Baraja la colección de cartas disponibles
     */
    private void shuffle() {
        Collections.shuffle(cards);
    }
    
    /**
     * @brief Comprueba si se han barajado las cartas
     * @return True si se han bajarado las cartas y False en otro caso
     */
    boolean justShuffled() {
        return (count==0);
    }
}

/*
class SafeHangarCardDeck extends CardDeck<Hangar> {
    @Override
    public Hangar next() {
        Hangar h=(Hangar)(super.next());
        return new Hangar(h) ;
    }
}
*/